### PR TITLE
WT-9950 __wt_realloc with clear memory flag requires knowledge of old size

### PR DESCRIPTION
**Pull request by mm-ng**

### CODE REVIEW

**Comment by mm-ng** [comment]
@ddanderson In the first commit we tried to replace the instances where `__wt_realloc` is called with a NULL `bytes_allocated_ret` argument with the `__wt_realloc_noclear` function instead. However, upon triggering the PR testing we found that there are other functions which indirectly call `__wt_realloc` and `__realloc_func`. An example is `wiredtiger_open` -> `__wt_scr_alloc_func` -> `__wt_realloc`. In these cases, sometimes the `bytes_allocated_ret` argument is NULL and sometimes it is not, meaning that we would need to explicitly check it every time to figure out whether we need to call `__wt_realloc_noclear` instead.

Given that the `__wt_realloc_noclear` function is only called once in the current code (in `__wt_buf_grow_worker`), we're thinking that we could probably get rid of this function and add an if/else check into `__wt_realloc` and call `__realloc_func` with clear_memory as true/false depending on whether `bytes_allocated_ret` is NULL.

**Comment by ruby-oujo** [comment]
Hi @ddanderson , we have currently changed `wt_realloc` in `curs_json.c` to `wt_realloc_noclear` as it is currently passing in a NULL for the second parameter. However, as you can see in the evergreen testing it is now failing the `test_jsondump01.py`. We have tried passing in the size variable and using `wt_realloc` but believe the variable is also NULL and we do not have access to another size variable. Is there an alternative way around this?
Would  there instances where we have a NULL `bytes_allocated_ret` but we still want to clear the memory?

**Comment by ddanderson** [comment]
@ruby-oujo @mm-ng , I think for cur_json, it must depend on the new memory getting cleared.  I think the easiest way is to use `__wt_realloc_def`.  You'll to add two size_t fields to WT_CURSOR_JSON, one for key one for value, to hold the previously allocated amount.

Another way (but I think it's less clear), is to have stack variable
```
 size_t mem_size;
  mem_size = 0; 
```
And pass in the address of this.  That will need a comment to explain.  I'd prefer the other approach, it's the normal way, so easy to understand quickly.

**Comment by mm-ng** [comment]
> __wt_realloc_def

@ddanderson We don’t quite understand what the previously allocated amount should be in the WT_CURSOR_JSON scenario. We initially thought that you meant we could store the size inside the __wt_json_to_item function, as this is used by the dump cursor to set the value when using JSON. However, when running the Python test, we see that this function is never entered. Is there somewhere else we should be saving the size inside the WT_CURSOR_JSON?

**Comment by ddanderson** [comment]
@mm-ng @ruby-oujo, the value of the new fields is zero when the WT_CURSOR_JSON is allocated, and the only place it needs to be managed is in `__wt_json_alloc_unpack`.  This is the cool thing about realloc/wt_realloc_def, it does all the space management for you.  Where it used to call:
```
  WT_RET(__wt_realloc(session, NULL, needed + 1, json_bufp));
```
it would call 
```
  WT_RET(__wt_realloc_def(session, &json->key_allocated, needed + 1, json_bufp));
```
in the key case, and `&json->value_allocated` in the value case - those are the two new fields.  You could make a variable to hold the address of` json->key_allocated` or `json->value_allocated`, as the case may be, so you still only have one `if` statement in the function.

As other examples, search the source tree for `list_allocated` or `ends_allocated`, there are many more.

**Comment by mm-ng** [comment]
> @mm-ng @ruby-oujo, the value of the new fields is zero when the WT_CURSOR_JSON is allocated, and the only place it needs to be managed is in `__wt_json_alloc_unpack`. This is the cool thing about realloc/wt_realloc_def, it does all the space management for you. Where it used to call:
> 
> ```
>   WT_RET(__wt_realloc(session, NULL, needed + 1, json_bufp));
> ```
> 
> it would call
> 
> ```
>   WT_RET(__wt_realloc_def(session, &json->key_allocated, needed + 1, json_bufp));
> ```
> 
> in the key case, and `&json->value_allocated` in the value case - those are the two new fields. You could make a variable to hold the address of` json->key_allocated` or `json->value_allocated`, as the case may be, so you still only have one `if` statement in the function.
> 
> As other examples, search the source tree for `list_allocated` or `ends_allocated`, there are many more.

Thanks for the clarification Don. We initially tried something like you suggested and were failing the `test_jsondump01` python test which was hitting an assertion in `__realloc_func`. It seems like the assertion is just catching a state that we are now allowing so we have changed the assertion statement to encompass this addition.

File: src/cursor/cur_json.c [hunk 1]
```diff
@@ -267,19 +267,21 @@ __wt_json_alloc_unpack(WT_SESSION_IMPL *session, const void *buffer, size_t size
   WT_CURSOR_JSON *json, bool iskey, va_list ap)
 {
     WT_CONFIG_ITEM *names;
-    size_t needed;
+    size_t bytes_allocated, needed;
```
Comment (ddanderson) [inline src/cursor/cur_json.c:270]: As coded, the call to `__wt_realloc_def` will update `bytes_allocated` (a stack variable), and json->value_size (or key_size) will never be non-zero.  I think what is needed is `size_t *bytes_allocatedp`, and set `bytes_allocatedp = &json->key_size` and pass `bytes_allocatedp` directly to `__wt_realloc_def`.  Let me know if you want to talk through it.

**Review by ddanderson — COMMENTED** [review]
I think there's still an issue in the call to realloc in the json cursor.

**Comment by ddanderson** [comment]
> Thanks for the clarification @ddanderson. We initially tried something like you suggested and were failing the `test_jsondump01` python test which was hitting an assertion in `__realloc_func`. It seems like the assertion is just catching a state that we are now allowing so we have changed the assertion statement to encompass this addition.

@mm-ng @ruby-oujo You probably don't have to change the assertion, see the suggested change in the code review.

File: src/cursor/cur_json.c (hunk 1 above)
Comment (mm-ng) [reply src/cursor/cur_json.c:270]: > As coded, the call to `__wt_realloc_def` will update `bytes_allocated` (a stack variable), and json->value_size (or key_size) will never be non-zero. I think what is needed is `size_t *bytes_allocatedp`, and set `bytes_allocatedp = &json->key_size` and pass `bytes_allocatedp` directly to `__wt_realloc_def`. Let me know if you want to talk through it.
> 
> After you make this change, please investigate if you can change the assertion back to the way it was in the realloc function?

I tried passing in the pointer directly as you suggested but the Python test fails to run. It's not clear from the output what's going wrong.

**Comment by sueloverso** [comment]
I don't know anything about this code so I'm removing myself as a reviewer.

**Review by ddanderson — APPROVED** [review]
LGTM!  Sorry for the runaround on the review, taking us in some other directions.
I opened WT-9986 to fix the problem, I'll own that - if you could just put a # FIXME-WT-9986 comment where the commented out lines in `test_jsondump01.py` are, that would be awesome.  Then ready to merge.

**Comment by mm-ng** [comment]
> LGTM! Sorry for the runaround on the review, taking us in some other directions. I opened [WT-9986](https://jira.mongodb.org/browse/WT-9986) to fix the problem, I'll own that - if you could just put a # FIXME-[WT-9986](https://jira.mongodb.org/browse/WT-9986) comment where the commented out lines in `test_jsondump01.py` are, that would be awesome. Then ready to merge.

Thanks Don! I'm going to hold off on merging for now because I've realised that `test_jsondump02.py` is also failing

**Comment by mm-ng** [comment]
> > LGTM! Sorry for the runaround on the review, taking us in some other directions. I opened [WT-9986](https://jira.mongodb.org/browse/WT-9986) to fix the problem, I'll own that - if you could just put a # FIXME-[WT-9986](https://jira.mongodb.org/browse/WT-9986) comment where the commented out lines in `test_jsondump01.py` are, that would be awesome. Then ready to merge.
> 
> Thanks Don! I'm going to hold off on merging for now because I've realised that `test_jsondump02.py` is also failing

@ddanderson  It looks like the failures in `test_jsondump02.py` could be related to the bug in the JSON cursor as well. Would you like us to disable this test as well as part of the changes?

**Comment by ddanderson** [comment]
@mm-ng @ruby-oujo, yes comment out or put a `self.skipTest` call in test_jsondump02.py, with a FIXME comment to the same ticket.

**Pull request merged**